### ISSUE TITLE
GuardedBy check fails with lambda expression

### ISSUE THREAD

**Issue by ben-manes** [google/error-prone#536]
Upgrading from 2.0.15 to 17 fails with the error,

```
BoundedLocalCache.java:1342: error: [GuardedByChecker] This access should be guarded by 'this.evictionLock', which is not currently held
      data.values().forEach(node -> removeNode(node, now))
```

The code it is complaining about is simplified to,
```
public void clear() {
  evictionLock.lock();
  try {
    ...
    data.values().forEach(node -> removeNode(node, now));
  } finally {
    evictionLock.unlock();
  }
}

@GuardedBy("evictionLock")
void removeNode(Node<K, V> node, long now) { ... }
```

Switching to a classic for-each loop resolves the problem.

**Comment by JakeWharton** [comment]
That code could just as easily be `threadPoolExecutor.execute(() -> removeNode(node, now))` where the failure would be warranted.

**Comment by ben-manes** [comment]
That's true, but a static analyzer would hopefully be able to detect that it is not crossing a thread boundary. Since it is not calling into an opaque interface, this seems possible.

**Comment by JakeWharton** [comment]
True! I just worry about how far you have to go here

**Comment by Stephan202** [comment]
@ben-manes beat me to it by 9 hours I see. I planned to create a minimal reproduction case for this same issue, but should have checked the issue tracker when I got up this morning. Anyway, in the process I did identify another issue, for which I filed #541.

NB: Note that if the lambda expression can be replaced with a method handle on the object under guard, then that will work too. (See the example code in the other issue.)

**Comment by JakeWharton** [comment]
Are you saying when the failing lambda is replaced with a method ref the
check no longer fails? That seems like a separate bug since they should
behave the same.

**Comment by Stephan202** [comment]
@JakeWharton yes, that's what I mean, though note that this does not apply to @ben-manes' case, because he uses a separate lock which is not "part" of the lambda expression. The linked issue #541 has a complete example of what I mean: the second test case uses method handles and passes the check, while the third test method uses a lambda expression and fails.

I agree that method handles should be considered (un)safe in the same contexts as lambda expressions. I'll file another ticket.

**Comment by Stephan202** [comment]
I filed #544. I hope the explanation is clear enough.

**Comment by graememorgan** [comment]
Done with the hardcoded list of `INVOKES_LAMBDAS_IMMEDIATELY` in the check.
